File: runtime/src/de/fatox/meta/ui/UiControlHelper.java
package de.fatox.meta.ui;

import com.badlogic.gdx.Input;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.Group;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.Button;
import com.badlogic.gdx.utils.Array;
import de.fatox.meta.Meta;
import de.fatox.meta.api.ui.UIManager;
import de.fatox.meta.injection.Inject;
import de.fatox.meta.injection.Singleton;
import de.fatox.meta.input.KeyListener;
import de.fatox.meta.input.MetaInput;

import java.util.Iterator;

@Singleton
public class UiControlHelper {
    @Inject
    private MetaInput metaInput;
    @Inject
    private UIManager uiManager;

    private Actor selectedActor = null;
    private Color selectedColor = Color.WHITE.cpy();
    private boolean activated = true;

    public UiControlHelper() {
        Meta.inject(this);
        metaInput.registerGlobalKeyListener(Input.Keys.RIGHT, new KeyListener() {
            @Override
            public void onEvent() {
                if (activated) {
                    setSelectedActor(getNextX(false));
                }
            }
        });
        metaInput.registerGlobalKeyListener(Input.Keys.LEFT, new KeyListener() {
            @Override
            public void onEvent() {
                if (activated) {
                    setSelectedActor(getNextX(true));
                }
            }
        });
        metaInput.registerGlobalKeyListener(Input.Keys.DOWN, new KeyListener() {
            @Override
            public void onEvent() {
                if (activated) {
                    setSelectedActor(getNextY(false));
                }
            }
        });
        metaInput.registerGlobalKeyListener(Input.Keys.UP, new KeyListener() {
            @Override
            public void onEvent() {
                if (activated) {
                    setSelectedActor(getNextY(true));
                }
            }
        });
        metaInput.registerGlobalKeyListener(Input.Keys.ENTER, new KeyListener() {
            @Override
            public void onEvent() {
                if (activated) {
                    InputEvent inputEvent = new InputEvent();
                    inputEvent.setButton(Input.Buttons.LEFT);
                    inputEvent.setStageX(selectedActor.getX());
                    inputEvent.setStageY(selectedActor.getY());
                    inputEvent.setType(InputEvent.Type.touchDown);
                    inputEvent.setListenerActor(selectedActor);
                    selectedActor.getListeners().forEach((eventListener -> eventListener.handle(inputEvent)));
                    inputEvent.setType(InputEvent.Type.touchUp);
                    selectedActor.getListeners().forEach((eventListener -> eventListener.handle(inputEvent)));
                }
            }
        });
    }

    private void targetsInGroup(Group t) {
        for (Actor actor : t.getChildren()) {
            if (actor instanceof Button && !targets.contains(actor, true)) {
                targets.add(actor);
            } else if (actor instanceof Group) {
                targetsInGroup((Group) actor);
            }
        }
    }

    Array<Actor> targets = new Array<>();

    private Array<Actor> getPossibleTargets() {
        targets.clear();
        if (!selectedActor.isVisible()) {
            selectedActor = selectedActor.getStage().getActors().get(0);
        }
        Group parent = selectedActor instanceof Group ? (Group) selectedActor : selectedActor.getParent();
        while (parent != null) {
            targetsInGroup(parent);
            parent = parent.getParent();
        }
        return targets;
    }

    private Vector2 helper = new Vector2();

    private Actor getNextX(boolean left) {
        Array<Actor> possibleTargets = getPossibleTargets();

        possibleTargets.sort((a1, a2) -> {
                    float a1x = a1.localToStageCoordinates(helper).x;
                    float a2x = a2.localToStageCoordinates(helper).x;
                    return (int) (a2x - a1x);
                }
        );
        Iterator<Actor> iterator = possibleTargets.iterator();
        while (iterator.hasNext()) {
            Actor next = iterator.next();
            if (Math.abs(next.getY() - selectedActor.getY()) > (selectedActor.getHeight() * 2.75f)) {
                iterator.remove();
            }
        }
        int index = possibleTargets.indexOf(selectedActor, true);
        return getNext(left, possibleTargets, index);
    }

    private Actor getNextY(boolean up) {
        Array<Actor> possibleTargets = getPossibleTargets();

        possibleTargets.sort((a1, a2) -> {
                    float a1y = a1.localToStageCoordinates(helper).y;
                    float a2y = a2.localToStageCoordinates(helper).y;
                    return (int) (a2y - a1y);
                }
        );
        Iterator<Actor> iterator = possibleTargets.iterator();
        while (iterator.hasNext()) {
            Actor next = iterator.next();
            if (Math.abs(next.getX() - selectedActor.getX()) > (selectedActor.getWidth() * 1.25f)) {
                iterator.remove();
            }
        }
        int index = possibleTargets.indexOf(selectedActor, true);
        return getNext(up, possibleTargets, index);
    }

    private Actor getNext(boolean left, Array<Actor> possibleTargets, int index) {
        if (left) {
            if (index == 0) {
                return possibleTargets.get(possibleTargets.size - 1);
            } else {
                return possibleTargets.get(index - 1);
            }
        } else {
            if (index == possibleTargets.size - 1) {
                return possibleTargets.get(0);
            } else {
                return possibleTargets.get(index + 1);
            }
        }
    }

    public Actor getSelectedActor() {
        return selectedActor;
    }

    public void setSelectedActor(Actor selectedActor) {
        if (this.selectedActor != null) {
            this.selectedActor.setColor(selectedColor);
        }
        this.selectedActor = selectedActor;
        this.selectedColor.set(selectedActor.getColor());
        this.selectedActor.setColor(Color.LIGHT_GRAY);
    }

}

File: runtime/src/de/fatox/meta/ui/UiControlHelper.kt
package de.fatox.meta.ui

import com.badlogic.gdx.Input
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.Group
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.Button
import com.badlogic.gdx.scenes.scene2d.utils.Disableable
import com.badlogic.gdx.utils.Array
import de.fatox.meta.api.MetaInputProcessor
import de.fatox.meta.api.registerGlobalKeyListener
import de.fatox.meta.injection.MetaInject.Companion.lazyInject
import kotlin.math.abs

object UiControlHelper {
	private val metaInput: MetaInputProcessor by lazyInject()

	var activated = true
	var canMove = true

	private var targets = Array<Actor>()

	private val selectedColor: Color = Color.WHITE.cpy()
	var selectedActor: Actor = Actor()
		set(value) {
			field.color = selectedColor // TODO The color of the old actor should be set here?
			field = value
			selectedColor.set(value.color)
			val color = field.color
			if (color.r + color.g + color.b > 1.0f)
				color.add(-0.25f, -0.25f, 0.3f, 0.2f)
			else
				color.add(0.05f, 0.05f, 0.3f, 0.2f)
		}

	private val possibleTargets: Array<Actor>
		get() {
			targets.clear()
			if (!selectedActor.isVisible) {
				selectedActor = selectedActor.stage.actors[0]
			}
			var parent: Group? = if (selectedActor is Group) selectedActor as Group else selectedActor.parent
			while (parent != null) {
				targetsInGroup(parent)
				parent = parent.parent
			}
			return targets
		}

	private val helper = Vector2()

	init {
		metaInput.registerGlobalKeyListener(Input.Keys.RIGHT) {
			if (activated && canMove)
				selectedActor = getNextX(left = false)
		}
		metaInput.registerGlobalKeyListener(Input.Keys.LEFT) {
			if (activated && canMove)
				selectedActor = getNextX(left = true)
		}
		metaInput.registerGlobalKeyListener(Input.Keys.DOWN) {
			if (activated && canMove)
				selectedActor = getNextY(up = false)
		}
		metaInput.registerGlobalKeyListener(Input.Keys.UP) {
			if (activated && canMove)
				selectedActor = getNextY(up = true)
		}
		metaInput.registerGlobalKeyListener(Input.Keys.ENTER) {
			if (activated) {
				val inputEvent = InputEvent().apply {
					button = Input.Buttons.LEFT
					stageX = selectedActor.x
					stageY = selectedActor.y
					type = InputEvent.Type.touchDown
					listenerActor = selectedActor
					stage = selectedActor.stage
				}
				selectedActor.listeners.forEach { eventListener -> eventListener.handle(inputEvent) }
				inputEvent.type = InputEvent.Type.touchUp
				selectedActor.listeners.forEach { eventListener -> eventListener.handle(inputEvent) }
			}
		}
	}

	private fun targetsInGroup(t: Group) {
		for (actor in t.children) {
			if (actor is Button && !targets.contains(actor, true)) {
				targets.add(actor)
			} else if (actor is Group) {
				targetsInGroup(actor)
			}
		}
	}

	private fun getNextX(left: Boolean): Actor {
		val possibleTargets = possibleTargets

		possibleTargets.sort { a1, a2 ->
			val a1x = a1.localToStageCoordinates(helper).x
			val a2x = a2.localToStageCoordinates(helper).x
			(a2x - a1x).toInt()
		}
		val iterator = possibleTargets.iterator()
		while (iterator.hasNext()) {
			val next = iterator.next()
			if (abs(next.y - selectedActor.y) > selectedActor.height * 2.75f
				|| (next is Disableable && next.isDisabled)) {
				iterator.remove()
			}
		}
		return getNext(left, possibleTargets, possibleTargets.indexOf(selectedActor, true))
	}

	private fun getNextY(up: Boolean): Actor {
		val possibleTargets = possibleTargets

		possibleTargets.sort { a1, a2 ->
			val a1y = a1.localToStageCoordinates(helper).y
			val a2y = a2.localToStageCoordinates(helper).y
			(a2y - a1y).toInt()
		}
		val iterator = possibleTargets.iterator()
		while (iterator.hasNext()) {
			val next = iterator.next()
			if (abs(next.x - selectedActor.x) > selectedActor.width * 1.25f
				|| (next is Disableable && next.isDisabled)) {
				iterator.remove()
			}
		}
		return getNext(up, possibleTargets, possibleTargets.indexOf(selectedActor, true))
	}

	private fun getNext(left: Boolean, possibleTargets: Array<Actor>, index: Int): Actor {
		return when {
			possibleTargets.isEmpty || index < 0 -> selectedActor
			left -> {
				if (index == 0)
					possibleTargets.get(possibleTargets.size - 1)
				else
					possibleTargets.get(index - 1)
			}
			index == possibleTargets.size - 1 -> possibleTargets.get(0)
			else -> possibleTargets.get(index + 1)
		}
	}
}
